x-pack/plugins/apm/public/utils/formatters.ts

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { memoize } from 'lodash';
+
+// tslint:disable-next-line  no-var-requires
+const numeral: (input: number) => Numeral = require('@elastic/numeral');
+
+interface Numeral {
+  format: (pattern: string) => string;
+}
+
+const UNIT_CUT_OFF = 10 * 1000000; // 10 seconds in microseconds
+
+export function asSeconds(value: number, withUnit: boolean = true) {
+  const formatted = asDecimal(value / 1000000);
+  return `${formatted}${withUnit ? ' s' : ''}`;
+}
+
+export function asMillis(value: number, withUnit: boolean = true) {
+  const formatted = asInteger(value / 1000);
+  return `${formatted}${withUnit ? ' ms' : ''}`;
+}
+
+export function asMillisWithDefault(value?: number) {
+  if (value == null) {
+    return `N/A`;
+  }
+  return asMillis(value);
+}
+
+export const getTimeFormatter: (
+  max: any
+) => (value: any, withUnit?: boolean) => string = memoize(
+  (max: any) => (max > UNIT_CUT_OFF ? asSeconds : asMillis)
+);
+
+export function timeUnit(max: any) {
+  return max > UNIT_CUT_OFF ? 's' : 'ms';
+}
+
+/*
+ * value: time in microseconds
+ */
+export function asTime(value: number): string {
+  return getTimeFormatter(value)(value);
+}
+
+export function asDecimal(value: number): string {
+  return numeral(value).format('0,0.0');
+}
+
+export function asInteger(value: number): string {
+  return numeral(value).format('0,0');
+}
+
+export function tpmUnit(type: string): string {
+  return type === 'request' ? 'rpm' : 'tpm';
+}
+
+export function asPercent(
+  numerator: number,
+  denominator: number = 0,
+  fallbackResult: any = ''
+): string {
```

Comment:
Right now the return type is `string` - but is that correct? The fallback value can be `any`. I think this function should always return a string. Defaulting to `''` is fine.

If not we should at least update the signature:
```ts
function asPercent<T>(
  numerator: number,
  denominator: number = 0,
  fallbackResult: T
): string | T {
```

Reply:
i agree that `asPercent` should always return `string`. if that changes, then we should ask why it's necessary.

Reply:
`fallbackResult` still accepts `any` so the following will be typed as `string` but will be a `boolean`:

```
const a = asPercent(0, 0, true);
```

Reply:
What if we just throw an error if the denominator is 0 and force ourselves to deal with that before using `asPercent`? I don't love doing "safe divide" logic inside `asPercent`, without being a safe divide function. But I agree that in almost all cases you want to deal with the "divide by zero" scenario before even calling these functions.

If we are nervous about throwing an error, then maybe at least force the fallback to be a string. In a general purpose "safeDivide" function, the fallback can be anything because you don't know how the caller will want to deal with a divide-by-zero, but in this case we know we want a percentage string so I think it's ok to force fallback to be a string.

Last thing: I saw somewhere in all this, someone was checking if a divide result was `=== Infinity` to see if the divide had failed. Just as an FYI, that's not sufficient because of this awesome thing:
```
0 / 0
NaN
```

Reply:
I feel we are overcomplicating this a bit. Instead of making a generic function that works for all conceivable use cases, let's just take a look at the use-cases we currently have. We can always adapt it later if it's the wrong fit.

We currently only use it twice:

https://github.com/elastic/kibana/blob/e51c2070efc209c8c0dcc108a66944aaadd0887b/x-pack/plugins/apm/public/components/app/TransactionDetails/Transaction/StickyTransactionProperties.tsx#L61

https://github.com/elastic/kibana/blob/e51c2070efc209c8c0dcc108a66944aaadd0887b/x-pack/plugins/apm/public/components/app/TransactionDetails/Transaction/WaterfallContainer/Waterfall/SpanFlyout/StickySpanProperties.tsx#L64

In the above cases we expect to get back `'X.XX%'` or `''` - and we don't have to wrap the statement in a try catch. Isn't that what we want?
We don't even need the fallback option currently, so I would be fine with removing it. But I'm always fine with leaving it as now.

Functionality-wise, I think it's good to go. We just need to fix the types because they are currently not correct.